src/dataframe/dataframe.jl

```diff
@@ -1741,78 +1769,96 @@ As a special case, if `df` has no columns and `row` is a `NamedTuple` or
 6×2 DataFrame
- Row │ A      B
-     │ Int64  Int64
-─────┼──────────────
-   1 │     1      1
-   2 │     2      2
-   3 │     3      3
-   4 │     1      0
-   5 │     1      1
-   6 │     1      0
+ Row │ A     B
+     │ Any   Int64
+─────┼─────────────
+   1 │ 11       12
+   2 │ a         1
+   3 │ true      0
+   4 │ b         2
+   5 │ c         3
+   6 │ a         1
 
 julia> push!(df, Dict(:A=>1.0, :C=>1.0), cols=:union)
 7×3 DataFrame
+ Row │ A     B        C
+     │ Any   Int64?   Float64?
+─────┼──────────────────────────
+   1 │ 11         12  missing
+   2 │ a           1  missing
+   3 │ true        0  missing
+   4 │ b           2  missing
+   5 │ c           3  missing
+   6 │ a           1  missing
+   7 │ 1.0   missing        1.0
+
+julia> insert!(df, 3, NamedTuple(), cols=:subset)
+8×3 DataFrame
  Row │ A        B        C
-     │ Float64  Int64?   Float64?
+     │ Any      Int64?   Float64?
 ─────┼─────────────────────────────
-   1 │     1.0        1  missing
-   2 │     2.0        2  missing
-   3 │     3.0        3  missing
-   4 │     1.0        0  missing
-   5 │     1.0        1  missing
-   6 │     1.0        0  missing
-   7 │     1.0  missing        1.0
-
-julia> push!(df, NamedTuple(), cols=:subset)
-8×3 DataFrame
- Row │ A          B        C
-     │ Float64?   Int64?   Float64?
-─────┼───────────────────────────────
-   1 │       1.0        1  missing
-   2 │       2.0        2  missing
-   3 │       3.0        3  missing
-   4 │       1.0        0  missing
-   5 │       1.0        1  missing
-   6 │       1.0        0  missing
-   7 │       1.0  missing        1.0
-   8 │ missing    missing  missing
+   1 │ 11            12  missing
+   2 │ a              1  missing
+   3 │ missing  missing  missing
+   4 │ true           0  missing
+   5 │ b              2  missing
+   6 │ c              3  missing
+   7 │ a              1  missing
+   8 │ 1.0      missing        1.0
 ```
 """
-function Base.push!(df::DataFrame, row::Any; promote::Bool=false)
+(push!, pushfirst!, insert!)
+
+Base.push!(df::DataFrame, row::Any; promote::Bool=false) =
+    insert!(df, nrow(df) + 1, row, promote=promote)
+
+Base.pushfirst!(df::DataFrame, row::Any; promote::Bool=false) =
+    insert!(df, 1, row, promote=promote)
+
+function Base.insert!(df::DataFrame, loc::Integer, row::Any; promote::Bool=false)
+    loc isa Bool && throw(ArgumentError("invalid index: $loc of type Bool"))
+    1 <= loc <= nrow(df)+1 || throw(ArgumentError("invalid index: $loc for data " *
+                                                "frame with $(nrow(df)) rows"))
     if !(row isa Union{Tuple, AbstractArray})
         # an explicit error is thrown as this was allowed in the past
-        throw(ArgumentError("`push!` does not allow passing collections of type " *
+        throw(ArgumentError("`insert!` does not allow passing collections of type " *
                             "$(typeof(row)) to be pushed into a DataFrame. Only " *
```

Comment:
Better not mention the function given that the error can happen when calling `push!` and `pushfirst!`?
```suggestion
        throw(ArgumentError("it is not allowed to insert collections of type " *
                            "$(typeof(row)) into a DataFrame. Only " *
```